File: src/main/scala/play/dev/filewatch/DefaultFileWatchService.scala
package play.dev.filewatch

import java.io.File
import java.nio.file.FileSystems

import io.methvin.watcher.DirectoryChangeEvent
import io.methvin.watcher.DirectoryChangeListener
import io.methvin.watcher.DirectoryWatcher
import io.methvin.watchservice.MacOSXListeningWatchService

import scala.collection.JavaConverters._
import scala.util.control.NonFatal

/**
 * Implementation of the file watch service that uses a native implementation for Mac and otherwise uses the JDK's
 * WatchService implementation.
 */
class DefaultFileWatchService(logger: LoggerProxy, isMac: Boolean, disableFileHashCheck: Boolean)
    extends FileWatchService {

  def this(logger: LoggerProxy) = this(logger, isMac = false, disableFileHashCheck = false)
  def this(logger: LoggerProxy, isMac: Boolean) = this(logger, isMac, disableFileHashCheck = false)

  def watch(filesToWatch: Seq[File], onChange: () => Unit) = {
    val dirsToWatch = filesToWatch.filter { file =>
      if (file.isDirectory) {
        true
      } else if (file.isFile) {
        logger.warn("An attempt has been made to watch the file: " + file.getCanonicalPath)
        logger.warn("DefaultFileWatchService only supports watching directories. The file will not be watched.")
        false
      } else false
    }

    val watchService = if (isMac) new MacOSXListeningWatchService() else FileSystems.getDefault.newWatchService()
    val directoryWatcher =
      DirectoryWatcher
        .builder()
        .paths(dirsToWatch.map(_.toPath).asJava)
        .listener(new DirectoryChangeListener {
          override def onEvent(event: DirectoryChangeEvent): Unit = onChange()
        })
        .fileHashing(!disableFileHashCheck)
        .watchService(watchService)
        .build()

    val thread = new Thread(
      new Runnable {
        override def run(): Unit = {
          try {
            directoryWatcher.watch()
          } catch {
            case NonFatal(_) => // Do nothing, this means the watch service has been closed, or we've been interrupted.
          }
        }
      },
      "play-watch-service"
    )
    thread.setDaemon(true)
    thread.start()

    new FileWatcher {
      override def stop(): Unit = directoryWatcher.close()
    }
  }
}

File: src/main/java/play/dev/filewatch/DefaultFileWatchService.java
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.dev.filewatch;

import io.methvin.watcher.DirectoryWatcher;
import io.methvin.watchservice.MacOSXListeningWatchService;
import java.io.File;
import java.io.IOException;
import java.nio.file.FileSystems;
import java.nio.file.Path;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

/**
 * Implementation of the file watch service that uses a native implementation for Mac and otherwise
 * uses the JDK's WatchService implementation.
 */
public class DefaultFileWatchService implements FileWatchService {

  private final LoggerProxy logger;
  private final boolean isMac;
  private final boolean disableFileHashCheck;

  public DefaultFileWatchService(LoggerProxy logger, boolean isMac, boolean disableFileHashCheck) {
    this.logger = logger;
    this.isMac = isMac;
    this.disableFileHashCheck = disableFileHashCheck;
  }

  public DefaultFileWatchService(LoggerProxy logger) {
    this(logger, false, false);
  }

  public DefaultFileWatchService(LoggerProxy logger, boolean isMac) {
    this(logger, isMac, false);
  }

  @Override
  public FileWatcher watch(Iterable<File> filesToWatch, Consumer<Optional<Path>> onChange) {
    var dirsToWatch =
        StreamSupport.stream(filesToWatch.spliterator(), false)
            .filter(
                file -> {
                  if (file.isDirectory()) {
                    return true;
                  } else if (file.isFile()) {
                    logger.warn(
                        () ->
                            "An attempt has been made to watch the file: "
                                + file.getAbsolutePath());
                    logger.warn(
                        () ->
                            "DefaultFileWatchService only supports watching directories. The file"
                                + " will not be watched.");
                    return false;
                  } else {
                    return false;
                  }
                });

    try {
      var watchService =
          isMac ? new MacOSXListeningWatchService() : FileSystems.getDefault().newWatchService();
      var directoryWatcher =
          DirectoryWatcher.builder()
              .paths(dirsToWatch.map(File::toPath).collect(Collectors.toList()))
              .listener(dirChangeEvent -> onChange.accept(Optional.of(dirChangeEvent.path())))
              .fileHashing(!disableFileHashCheck)
              .watchService(watchService)
              .build();

      var thread =
          new Thread(
              () -> {
                try {
                  directoryWatcher.watch();
                } catch (VirtualMachineError | ThreadDeath | LinkageError e) {
                  throw e;
                } catch (Throwable ignored) {
                  // Do nothing, this means the watch service has been closed, or we've been
                  // interrupted.
                }
              },
              "play-watch-service");
      thread.setDaemon(true);
      thread.start();

      return () -> {
        try {
          directoryWatcher.close();
        } catch (IOException e) {
          logger.error(e::getMessage);
        }
      };
    } catch (IOException e) {
      logger.error(e::getMessage);
      throw new RuntimeException(e.getMessage(), e);
    }
  }
}
